src/client/signatures/canonical_json/src/ser.rs

```diff
@@ -0,0 +1,595 @@
+            
+            writer.write(&"\\u".as_bytes())?;
+            string_iter.next();
+
+            if string_iter.peek() == Some(&'{') {
+                // consume at most 4 characters till '}' is found
+                let mut characters = String::new();
+                string_iter.next(); // skip the '{' for now
+                let mut index = 0;
+
+                while index < 4 && string_iter.peek() != Some(&'}') && string_iter.peek() != None {
+
+                    match string_iter.peek() {
+                        Some(character) => characters.push(*character),
+                        None => break
+                    };
+
+                    string_iter.next();
+                    index += 1;
+                }
+
+                if string_iter.peek() == None {
+                    // could not find '}' bracket so must include '{' and following characters
+                    writer.write(&"{".as_bytes())?;
+                    writer.write(&characters.into_bytes())?;
+                } else if string_iter.peek() == Some(&'}') {
+                    // found '}' - remove '{' and '}' but must pad zeros
+                    if characters.len() == 0 {
+                        writer.write(&"{}".as_bytes())?;
+                    } else {
+                        writer.write(&std::iter::repeat("0").take(4 - characters.len()).collect::<String>().into_bytes())?;
+                        writer.write(&characters.into_bytes())?;
+                        string_iter.next(); // skip '}' 
+                    }
+                }
+            }
+
+            continue;
+        }
+
+        writer.write(curr_char.to_string().as_bytes())?;
+    }
+
+    Ok(())
+}
+
+/// Serialize a JSON value to String
+/// 
+/// # Examples
+/// ```rust
+/// # use canonical_json::ser::to_canonical_json;
+/// # use serde_json::json;
+/// # fn main() {
+///     to_canonical_json(&json!(null)); // returns "null"
+/// 
+///     to_canonical_json(&json!("test")); // returns "test"
+/// 
+///     to_canonical_json(&json!(10.0_f64.powf(21.0))); // returns "1e+21"
+/// 
+///     to_canonical_json(&json!({
+///         "a": "a",
+///         "id": "1",
+///         "b": "b"
+///     })); // returns "{"a":"a","b":"b","id":"1"}"; (orders object keys)
+/// 
+///     to_canonical_json(&json!(vec!["one", "two", "three"])); // returns "["one","two","three"]"
+/// # }
+/// 
+/// ```
+pub fn to_canonical_json(input: &serde_json::Value) -> String {
+
+    let string = vec![];
+    let null_string = "null".to_owned();
+    let mut serializer = serde_json::Serializer::with_formatter(string, JSONFormatter{});
+    match input.serialize(&mut serializer) {
+        Ok(()) => match String::from_utf8(serializer.into_inner()) {
+            Ok(serialized_string) => return serialized_string,
+            Err(error) => {
+                error!("Error converting bytes to string: {}", error);
+                return null_string;
```

Comment:
This doesn't make much sense to me: why not return `Result<String, ...>` and let consumers decide if they want this `null_string` value?